### PR TITLE
fix(cli): make Escape stop only the focused agent

### LINKED ISSUE #9007: fix(cli): make Escape stop only the focused agent
## Summary

In the chat TUI, a bare `Esc` outside a modal currently stops the root chat run, irrespective of which agent stream is focused. Since the root stop normally cascades to its children, one `Esc` can stop every running agent.

`Esc` must instead stop exactly the running or waiting agent stream that was focused when the key was pressed. It must not terminate sibling agents or dispose the CLI process session. After a focused child is stopped, focus must return to its immediate owner so Up/Down navigation works again.

This is a user-visible correction and should receive a concise entry under the current Unreleased changelog section.

## Confirmed root cause

The TUI already owns the focused stream identity in `activeStreamId`, but the Escape contract carries only root-run callbacks. `runChatTui` maps that callback to `chatController.stop()`, which stops the root `session.streamId`; `ExecutionRegistry.stopAgentStream` then cascades through active children by default. The focused identity is discarded before cancellation reaches the runtime.

The root stop also invokes global approval cleanup, so a focused-stream gesture can settle approvals belonging to siblings.

A second state-machine defect causes the reported arrow-key failure. Stream-status handling returns a focused child to its owner only after a lifecycle terminal transition, not a user-stop transition. The cancelled child therefore remains focused; the composer remains disabled for that terminal child, and its Up/Down handlers are inert.

Historical #4288 and merged #4298 intentionally made Escape equivalent to root interruption before multi-stream focus existed. #7496 concerns Escape-chord timing, not cancellation ownership. Searches across open issues and PRs found no duplicate.

## Required invariant

- A modal, form, palette, reverse search, task detail, information pane, or child-list selection owns `Esc` while active; the key dismisses or backs out of that surface only.
- Otherwise, bare `Esc` captures the currently focused stream identity.
- If that stream is running, resuming, streaming, or waiting, `Esc` stops exactly that stream.
- The delayed bare-Escape path used to distinguish `Esc` from `Esc 1..9` retains the identity captured at the physical keypress. A later focus change cannot retarget the pending stop.
- Stopping `main` through `Esc` preserves active agent children by detaching or promoting them. Stopping a child preserves its parent, siblings, and active agent descendants.
- Processes owned internally by the stopped agent may terminate as part of stopping that agent.
- A child-only stop does not set the root TUI session’s `stopRequested`.
- Approval cancellation is scoped to the stopped stream. Pending sibling approvals remain pending and usable.
- A user-stopped attached child returns focus to its immediate owner. Unrelated or detached stream transitions do not move focus.
- After the stop settles, Up/Down immediately regain their ordinary composer-history or child-list navigation behavior.
- `Esc` never exits the TUI, disposes `SessionHandle`, or terminates the process session.

## Implementation boundary

Carry a `StreamTabId` through the Escape policy instead of reusing the root-only interrupt callback. Preserve the stream identity across the existing 500 ms Escape-chord delay.

Route the focused stop through the existing session controller and `ExecutionRegistry.stopAgentStream`; do not introduce a second cancellation registry or replacement facade. The Escape-local route forces preservation of other agent streams and uses stream-scoped interaction cancellation. The existing root/session stop path remains available to process-lifetime commands.

The verifier corrections above are hard requirements. No persisted or public wire format changes are intended.

## Non-goals

These retain their present broader semantics: `Ctrl-C`, `/exit`, process shutdown and signal teardown, explicit stop or kill commands, and configurable detach-on-stop policy outside this Escape-local action. This issue does not change `Esc 1..9` focus chords.

## Deterministic verification

- Focus child A, dispatch Escape, change focus to child B before the delayed callback, and prove the captured target remains A.
- Exercise every keyboard-owning surface and prove Escape only dismisses that surface.
- With root plus two active children, stop root through the Escape-local route and prove both children survive and the process session remains alive.
- Stop child A and prove root, sibling B, root `stopRequested`, and child A have the required independent states.
- Queue approvals for A and B; stop A and prove B remains pending and decidable.
- Prove a focused attached child returns to its owner after `user-stop`, while unrelated and detached transitions do not move focus.
- After stopping the focused child, send Up/Down and prove history recall and child-list navigation work.
- Add raw-byte PTY scenarios for main-focused, child-focused, modal-owned, and `Esc`-number chord cases.

Run formatting, type checking, `compile:fast`, lint, dead-code ratchet, focused Vitest suites, and the relevant PTY validator scenarios.

### PR DESCRIPTION
**Pull request by LionSR**
## Summary

- route bare Escape to the focused running or waiting stream instead of the root run
- preserve sibling agents, active descendants, sibling approvals, and the process session
- return a user-stopped child to its immediate owner so prompt history and child-list navigation work immediately
- retain the interaction host for detached children without retaining overlapping root-result presenters

## Design

The TUI captures the focused stream identifier at the physical Escape keypress and carries it through the existing Escape-number chord delay. Stream-scoped cancellation uses the existing execution registry with active agent children detached. Modal, palette, reverse-search, task-detail, information-pane, and child-list ownership remain unchanged.

Root completion and presentation lifetimes are distinct: the completed root row is finalized immediately; its terminal-result presenter is detached at root finalization; the interaction adapter and runtime host remain only while inherited child executions still need approvals. A subscribe-before-check liveness listener prevents lost cleanup wakeups.

## Verification

- 224 focused Vitest tests
- six raw PTY Escape scenarios: prompt history, root-only stop, child-only stop, captured target, modal ownership, and Escape-number focus
- npm run typecheck
- npm run compile:fast
- npm run lint (zero errors; one unrelated existing warning)
- npm run check:dead-code-ratchet
- npm run format
- git diff --check
- two independent adversarial reviews, including overlapping root generations and detached-child approvals

## Line accounting

- production source: +140 / -43, net +97
- test and PTY verification: +803 / -14, net +789
- documentation and changelog: +10 / -1, net +9
- total: +953 / -58, net +895

The increase is concentrated in deterministic concurrency and raw-terminal regression coverage. No persisted or public wire format changes are intended.

Closes #9007

---

> [!NOTE]
> **Medium Risk**
> Touches execution stop policy, session host/interaction handoff, and TUI focus routing—user-visible orchestration behavior with concurrency edge cases covered by new tests.
> 
> **Overview**
> **Escape** in the CLI TUI now stops the **focused** running or waiting agent stream instead of always interrupting the root run. **Ctrl-C** behavior for the root session is unchanged; modals, palettes, and other foreground UI still own Escape first.
> 
> Stream-scoped stop goes through new **`stopStream`** / **`onInterruptStream`** wiring, with the meta-chord delay carrying the stream id captured at keypress. Stopping a focused child returns navigation to its **owner** (same as lifecycle completion), so prompt history and child-list focus work right after stop.
> 
> Runtime hosting is split so a stopped root can **finalize its result toast** while **detached children** keep approvals and TUI facts: run-fact subscription is session-scoped, host interactions release only when no execution still uses that host, and switching interaction hosts **cancels** the previous adapter’s pending presentations instead of ignoring stale results.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 4781e679bd271f28070ec3f76a2160542c1565cb. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/cli/src/chat/chatSessionController.ts [hunk 1]
```diff
@@ -291,26 +303,66 @@ export function createChatSessionController(
...
-          session.runtimeHost = undefined;
-        }
+        // The root terminal result is published before its run promise
+        // settles. Children retain `isSubagent: true` and never produce a
+        // terminal toast, so this session-wide listener has no work after the
+        // root finalizes and must not overlap a later root's listener.
+        detachResultToastOnce();
         markChatTuiRunCompleted(session);
-        void runtimeHost.close();
+        releaseHostWhenUnused();
```
[the remark this answers is not in the document]
Comment (LionSR) [reply packages/cli/src/chat/chatSessionController.ts:365]: Addressed the useful documentation portion in c8f532538f: AppProps now distinguishes the Escape stream-scoped contract from the Ctrl-C root-run contract, and the setupRunHost comment distinguishes the root terminal presenter from the interactions retained for detached children. The release closures remain local intentionally: they share one captured lifecycle state and extracting them would add a single-consumer abstraction contrary to the repository guidance. The harness comments remain consistent with these lifetimes.

**Comment by LionSR** [comment]
This is a little surprising—a lot of code has been added. Let's try to see if it's possible to work with a simpler system, so there's less chance for bugs and fewer components, each with clear ownership.

**Comment by LionSR** [comment]
continue and rebase this

File: packages/cli/src/chat/chatSessionController.ts [hunk 2]
```diff
@@ -185,6 +188,13 @@ export function createChatSessionController(
   let interruptedContinuation: InterruptedContinuationBatch | undefined;
   let pendingInterruptedFollowUps: InterruptedFollowUp[] = [];
 
+  // Run facts belong to the TUI session, not to any one root turn. A stopped
+  // root may leave detached children running, and those children must keep
+  // projecting status, output, and approval-related facts after the root
+  // promise settles. Installing this once also avoids duplicate projections
+  // when another root starts while an earlier detached child is still alive.
+  disposers.push(attachTuiRunFactSubscription(defaultSession().events));
```
[the remark this answers is not in the document]
Comment (LionSR) [reply packages/cli/src/chat/chatSessionController.ts:196]: Fixed in 4781e679bd. The session-scoped run-fact subscription now registers with the existing CLI reset-hook boundary and advances its captured generation after every reset. A deterministic test resets state, emits facts for the next root, and verifies they project without reinstalling the subscription.

File: packages/cli/src/chat/chatSessionController.ts [hunk 3]
```diff
@@ -291,26 +303,66 @@ export function createChatSessionController(
...
+      detachResultToast();
+    };
+    let released = false;
+    let detachExecutionListener = (): void => {};
+    const releaseHost = (): void => {
+      if (released) return;
+      released = true;
+      detachExecutionListener();
+      detachResultToastOnce();
+      detachHostInteractions();
```
[the remark this answers is not in the document]
Comment (LionSR) [reply packages/cli/src/chat/chatSessionController.ts:323]: Fixed in 4781e679bd at the session interaction owner. Before an attachment handoff redispatches a pending request, SessionHostInteractions now cancels the previous adapter presentation after invalidating its dispatch version. The regression test covers A → B → A and proves each stale presentation is gone before the next one appears.

File: packages/cli/src/chat/chatSessionController.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (LionSR) [reply packages/cli/src/chat/chatSessionController.ts:196]: Fixed in 4781e679bd by the same reset-generation repair as the Codex review. The once-per-session subscription now advances with resetCliState instead of becoming permanently stale, with direct post-reset fact coverage.

File: packages/cli/scripts/tui-harness.tsx [hunk 4]
```diff
@@ -2072,12 +2155,13 @@ const ink = render(renderHarnessApp(), {
 });
 inkRef.current = ink;
 
+CHILD_EVENT_ORDER_DISPOSERS.push(subscribeStreamStatus());
+
```
[the remark this answers is not in the document]
Comment (LionSR) [reply packages/cli/scripts/tui-harness.tsx:2154]: Fixed in 4781e679bd. The harness comment now matches the once-per-TUI-session lifetime, and canInterruptHarnessStream is pure; the delayed focus retarget belongs to the interruption action instead. All six real PTY Escape scenarios pass.
